docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + 41 + 17
+```
+
+The rewrite rules then detects that `1 + 41` is `42`:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  42 + 17
+```
+
+The rewrite rules then detects that `42 + 17` is `59`:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  59
+```
+
+Lastly, `purs-backend-es` determines that the `a` binding is never used; thus, it's removed entirely:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = 59
+```
+
+Now that we've finished optimizing `foo`, the following states are true:
+
+- `ignoreArgs` is still defined as it originally was
+- the optimized `foo` still represents the same value as the unoptimized `foo` would have had at runtime.
+
+## A Methodology for Defining Inline Directives
+
+Inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes. Ideally, inlining values will always trigger rewrite rules that both reduce the bundle size of the code AND make the resulting program more performant. However, inlining may increase a program's bundle size without improving its performance.
+
+In short, this process isn't scientific. While one can add inline directives without much thought, the result likely won't be what they want.
+
+So, how should one determine if an inline directive needs to be added? One should always use the below methodology:
+
+1. Think of a snippet of code you want to ensure is optimized.
+2. Define a snapshot for that snippet.
+3. Look at the current JavaScript output of that snapshot.
+4. If the current output is not good enough,
+    1. add an inline directive to the outermost thing
+       1. When in doubt, use `arity=x` where `x` is the number of args the function in the outputted JavaScript, not in PureScript source code, takes.
+    2. Go to Step 3.
+5. Once the output satisfies you, determine where it should be defined (see next section)
+
+First, think of a snippet of code you want to ensure is optimized. If you don't have a goal in mind, you will add inline directives that will unnecessarily bloat your code.
+
+Second, define a snapshot for that snippet. Without a small snippet of code, you won't be able to see what affects adding more directives may have.
+
+Third, look at the current JavaScript output of that snapshot. If it's already as optimized, there's no reason to add an inline directive. You're done.
+
+Otherwise, fourth, find the outermost function and add an inline directive for that. Because `purs-backend-es` operates on `CoreFn`, not PureScript source code, refer to the function in the outputted JavaScript to determine how many args the function takes.
```

Comment:
Is this just repeating the steps above?

Reply:
Yes, but each expands on the heading and provides more context, except... I thought there would be more to say for each step, but apparently not.